roachtest: kv50/enc=false/nodes=4/cpu=96/batch=64 failed

SHA: https://github.com/cockroachdb/cockroach/commits/7ce9188c6e64465d9dcb9f0ca0f113dd0e584da0

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=kv50/enc=false/nodes=4/cpu=96/batch=64 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1178908&tab=buildLog

```
The test failed on release-2.1:
	cluster.go:1267,kv.go:69,cluster.go:1605,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1178908-kv50-enc-false-nodes-4-cpu-96-batch-64:5 -- ./workload run kv --init --splits=1000 --histograms=logs/stats.json --concurrency=256 --duration=10m --read-percent=50 --batch=64 {pgurl:1-4} returned:
		stderr:
		
		stdout:
		 read
		     44s        0          232.9          194.4    671.1   2080.4   2550.1   3489.7 write
		     45s        0          210.1          201.8    167.8    838.9   1208.0   1610.6 read
		     45s        0          237.1          195.4    637.5   1811.9   2952.8   4026.5 write
		     46s        0          216.0          202.1    285.2    872.4   1275.1   1744.8 read
		     46s        0          261.9          196.8    637.5   2147.5   3221.2   5368.7 write
		     47s        0          219.1          202.4    302.0    805.3   1275.1   1610.6 read
		     47s        0          227.1          197.4    704.6   1677.7   3087.0   4563.4 write
		     48s        0          121.0          200.7    260.0    838.9    939.5   1040.2 read
		     48s        0          135.0          196.1    771.8   1744.8   2147.5   2684.4 write
		     49s        0            0.0          196.7      0.0      0.0      0.0      0.0 read
		     49s        0            0.0          192.1      0.0      0.0      0.0      0.0 write
		Error: EOF
		Error:  exit status 1
		: exit status 1
	cluster.go:1626,kv.go:72,kv.go:137,test.go:1214: unexpected node event: 3: dead

```

Uh oh. @nvanbenschoten take a look at this. Are we maybe missing some backport for the data races you recently fixed? All failures are on 2.1.

```
E190421 10:37:11.545209 141697 util/log/crash_reporting.go:203  [n3,client=10.142.0.133:46312,user=root,txn=57ea2cc6] a panic has occurred!
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 141697 [running]:
panic(0x2777680, 0x4260d80)
	/usr/local/go/src/runtime/panic.go:551 +0x3c1 fp=0xc42fd7e318 sp=0xc42fd7e278 pc=0x6de8e1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4208f8090, 0x305d860, 0xc42be55890)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:184 +0x11f fp=0xc42fd7e378 sp=0xc42fd7e318 pc=0xf0f58f
runtime.call32(0x0, 0x2c3c328, 0xc428bf63f0, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:573 +0x3b fp=0xc42fd7e3a8 sp=0xc42fd7e378 pc=0x70bc7b
panic(0x2777680, 0x4260d80)
	/usr/local/go/src/runtime/panic.go:502 +0x229 fp=0xc42fd7e448 sp=0xc42fd7e3a8 pc=0x6de749
runtime.panicindex()
	/usr/local/go/src/runtime/panic.go:28 +0x5e fp=0xc42fd7e468 sp=0xc42fd7e448 pc=0x6dd58e
github.com/cockroachdb/cockroach/pkg/roachpb.(*Value).MarshalTo(0xc42d197378, 0xc42fcb7949, 0xa, 0xa, 0xc, 0x14, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/data.pb.go:1069 +0x21d fp=0xc42fd7e4c8 sp=0xc42fd7e468 pc=0xe6a17d
github.com/cockroachdb/cockroach/pkg/roachpb.(*PutRequest).MarshalTo(0xc42d197340, 0xc42fcb7935, 0x1e, 0x1e, 0x20, 0x2, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:6905 +0x1ba fp=0xc42fd7e528 sp=0xc42fd7e4c8 pc=0xdce06a
github.com/cockroachdb/cockroach/pkg/roachpb.(*RequestUnion_Put).MarshalTo(0xc42848a358, 0xc42fcb7933, 0x20, 0x20, 0xdf1409, 0xc42848a358, 0x22)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:10346 +0xe1 fp=0xc42fd7e578 sp=0xc42fd7e528 pc=0xde27e1
github.com/cockroachdb/cockroach/pkg/roachpb.(*RequestUnion).MarshalTo(0xc42fd7e620, 0xc42fcb7933, 0x20, 0x20, 0x22, 0xb3, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:10317 +0x73 fp=0xc42fd7e5c0 sp=0xc42fd7e578 pc=0xde2563
github.com/cockroachdb/cockroach/pkg/roachpb.(*BatchRequest).MarshalTo(0xc434875100, 0xc42fcb7880, 0xd3, 0xd3, 0xd3, 0xd3, 0x2aa8a00)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:11729 +0x245 fp=0xc42fd7e640 sp=0xc42fd7e5c0 pc=0xdea4f5
github.com/cockroachdb/cockroach/pkg/roachpb.(*BatchRequest).Marshal(0xc434875100, 0x2aa8a00, 0xc434875100, 0x7fe5af4a2650, 0xc434875100, 0xc42fd7e601)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:11704 +0x7f fp=0xc42fd7e698 sp=0xc42fd7e640 pc=0xdea22f
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/encoding/proto.codec.Marshal(0x2aa8a00, 0xc434875100, 0x6b6f16, 0x3, 0xc4200cc070, 0xc4200cc000, 0xc434998f58)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/encoding/proto/proto.go:70 +0x19e fp=0xc42fd7e718 sp=0xc42fd7e698 pc=0xbd60be
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/encoding/proto.(*codec).Marshal(0x48833e8, 0x2aa8a00, 0xc434875100, 0x6de194, 0x2c3e220, 0xc42fd7e7d0, 0x10, 0xc42c768900)
	<autogenerated>:1 +0x48 fp=0xc42fd7e760 sp=0xc42fd7e718 pc=0xbd6838
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.encode(0x7fe5af43e678, 0x48833e8, 0x2aa8a00, 0xc434875100, 0xc429f423f0, 0x0, 0xc429f423f0, 0x0, 0xbf272fd5e07dffd2)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/rpc_util.go:487 +0x61 fp=0xc42fd7e7e8 sp=0xc42fd7e760 pc=0xbee9e1
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*csAttempt).sendMsg(0xc429d47ad0, 0x2aa8a00, 0xc434875100, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/stream.go:479 +0xc9 fp=0xc42fd7e908 sp=0xc42fd7e7e8 pc=0xbfbb99
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*clientStream).SendMsg(0xc434875200, 0x2aa8a00, 0xc434875100, 0xc420a3cf00, 0x2b1f443)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/stream.go:400 +0x43 fp=0xc42fd7e940 sp=0xc42fd7e908 pc=0xbfb763
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.invoke(0x305d860, 0xc42be55890, 0x2b1f443, 0x21, 0x2aa8a00, 0xc434875100, 0x2997ca0, 0xc420ef1d40, 0xc420a3cf00, 0xc4259d5200, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/call.go:75 +0xfe fp=0xc42fd7e9c8 sp=0xc42fd7e940 pc=0xbde54e
github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingClientInterceptor.func1(0x305d860, 0xc42be55890, 0x2b1f443, 0x21, 0x2aa8a00, 0xc434875100, 0x2997ca0, 0xc420ef1d40, 0xc420a3cf00, 0x2c3e210, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/client.go:44 +0xe28 fp=0xc42fd7ede8 sp=0xc42fd7e9c8 pc=0x14d1268
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*ClientConn).Invoke(0xc420a3cf00, 0x305d860, 0xc42be55890, 0x2b1f443, 0x21, 0x2aa8a00, 0xc434875100, 0x2997ca0, 0xc420ef1d40, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/call.go:35 +0x109 fp=0xc42fd7eea0 sp=0xc42fd7ede8 pc=0xbde109
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.Invoke(0x305d860, 0xc42be55890, 0x2b1f443, 0x21, 0x2aa8a00, 0xc434875100, 0x2997ca0, 0xc420ef1d40, 0xc420a3cf00, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/call.go:60 +0xc1 fp=0xc42fd7ef20 sp=0xc42fd7eea0 pc=0xbde421
github.com/cockroachdb/cockroach/pkg/roachpb.(*internalClient).Batch(0xc42592c7f8, 0x305d860, 0xc42be55890, 0xc434875100, 0x0, 0x0, 0x0, 0x305d860, 0xc42be55890, 0x100000002)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:6570 +0xd2 fp=0xc42fd7efa8 sp=0xc42fd7ef20 pc=0xdcc0d2
github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).sendBatch(0xc429f42390, 0x305d860, 0xc42be55890, 0x303a360, 0xc42592c7f8, 0x0, 0x0, 0x20000
```

--- a/pkg/internal/client/txn_test.go
+++ b/pkg/internal/client/txn_test.go
@@ -392,7 +392,6 @@ func TestSetPriority(t *testing.T) {
 				}
 
 				br := &roachpb.BatchResponse{}
-				br.Txn = &roachpb.Transaction{}
 				br.Txn.Update(ba.Txn) // copy
 				return br, nil
 			}), clock)

--- a/pkg/kv/dist_sender.go
+++ b/pkg/kv/dist_sender.go
@@ -935,7 +935,9 @@ func (ds *DistSender) divideAndSendBatchToRanges(
 			// Update the transaction from the response. Note that this wouldn't happen
 			// on the asynchronous path, but if we have newer information it's good to
 			// use it.
-			ba.UpdateTxn(resp.reply.Txn)
+			if !lastRange {
+				ba.UpdateTxn(resp.reply.Txn)
+			}
 
 			mightStopEarly := ba.MaxSpanRequestKeys > 0 || stopAtRangeBoundary
 			// Check whether we've received enough responses to exit query loop.

--- a/pkg/roachpb/api.go
+++ b/pkg/roachpb/api.go
@@ -383,10 +383,13 @@ func (h *BatchResponse_Header) combine(o BatchResponse_Header) error {
 		)
 	}
 	h.Timestamp.Forward(o.Timestamp)
-	if h.Txn == nil {
-		h.Txn = o.Txn
-	} else {
-		h.Txn.Update(o.Txn)
+	if txn := o.Txn; txn != nil {
+		if h.Txn == nil {
+			txnClone := txn.Clone()
+			h.Txn = &txnClone
+		} else {
+			h.Txn.Update(txn)
+		}
 	}
 	h.Now.Forward(o.Now)
 	h.CollectedSpans = append(h.CollectedSpans, o.CollectedSpans...)

--- a/pkg/roachpb/batch.go
+++ b/pkg/roachpb/batch.go
@@ -60,17 +60,17 @@ func (ba *BatchRequest) SetActiveTimestamp(nowFn func() hlc.Timestamp) error {
 // UpdateTxn updates the batch transaction from the supplied one in
 // a copy-on-write fashion, i.e. without mutating an existing
 // Transaction struct.
-func (ba *BatchRequest) UpdateTxn(otherTxn *Transaction) {
-	if otherTxn == nil {
+func (ba *BatchRequest) UpdateTxn(o *Transaction) {
+	if o == nil {
 		return
 	}
-	otherTxn.AssertInitialized(context.TODO())
+	o.AssertInitialized(context.TODO())
 	if ba.Txn == nil {
-		ba.Txn = otherTxn
+		ba.Txn = o
 		return
 	}
 	clonedTxn := ba.Txn.Clone()
-	clonedTxn.Update(otherTxn)
+	clonedTxn.Update(o)
 	ba.Txn = &clonedTxn
 }
 

--- a/pkg/roachpb/batch.go
+++ b/pkg/roachpb/batch.go
@@ -365,7 +365,6 @@ func (br *BatchResponse) Combine(otherBatch *BatchResponse, positions []int) err
 			return errors.Errorf("can not combine %T and %T", valLeft, valRight)
 		}
 	}
-	br.Txn.Update(otherBatch.Txn)
 	return nil
 }
 

--- a/pkg/roachpb/data.go
+++ b/pkg/roachpb/data.go
@@ -770,18 +770,10 @@ func (t Transaction) LastActive() hlc.Timestamp {
 	return ts
 }
 
-// Clone creates a copy of the given transaction. The copy is "mostly" deep,
-// but does share pieces of memory with the original such as Key, ID and the
-// keys with the intent spans.
+// Clone creates a copy of the given transaction. The copy is shallow because
+// none of the references held by a transaction allow interior mutability.
+// TODO(nvanbenschoten): Clean this up in the next commit.
 func (t Transaction) Clone() Transaction {
-	mt := t.ObservedTimestamps
-	if mt != nil {
-		t.ObservedTimestamps = make([]ObservedTimestamp, len(mt))
-		copy(t.ObservedTimestamps, mt)
-	}
-	// Note that we're not cloning the span keys under the assumption that the
-	// keys themselves are not mutable.
-	t.Intents = append([]Span(nil), t.Intents...)
 	return t
 }
 
@@ -1667,7 +1659,8 @@ func (kv KeyValueByKey) Swap(i, j int) {
 
 var _ sort.Interface = KeyValueByKey{}
 
-// observedTimestampSlice maintains a sorted list of observed timestamps.
+// observedTimestampSlice maintains an immutable sorted list of observed
+// timestamps.
 type observedTimestampSlice []ObservedTimestamp
 
 func (s observedTimestampSlice) index(nodeID NodeID) int {
@@ -1689,19 +1682,27 @@ func (s observedTimestampSlice) get(nodeID NodeID) (hlc.Timestamp, bool) {
 }
 
 // update the timestamp for the specified node, or add a new entry in the
-// correct (sorted) location.
+// correct (sorted) location. The receiver is not mutated.
 func (s observedTimestampSlice) update(
 	nodeID NodeID, timestamp hlc.Timestamp,
 ) observedTimestampSlice {
 	i := s.index(nodeID)
 	if i < len(s) && s[i].NodeID == nodeID {
 		if timestamp.Less(s[i].Timestamp) {
-			s[i].Timestamp = timestamp
+			// The input slice is immutable, so copy and update.
+			cpy := make(observedTimestampSlice, len(s))
+			copy(cpy, s)
+			cpy[i].Timestamp = timestamp
+			return cpy
 		}
 		return s
 	}
-	s = append(s, ObservedTimestamp{})
-	copy(s[i+1:], s[i:])
-	s[i] = ObservedTimestamp{NodeID: nodeID, Timestamp: timestamp}
-	return s
+	// The input slice is immutable, so copy and update. Don't append to
+	// avoid an allocation. Doing so could invalidate a previous update
+	// to this receiver.
+	cpy := make(observedTimestampSlice, len(s)+1)
+	copy(cpy[:i], s[:i])
+	cpy[i] = ObservedTimestamp{NodeID: nodeID, Timestamp: timestamp}
+	copy(cpy[i+1:], s[i:])
+	return cpy
 }

--- a/pkg/roachpb/data.pb.go
+++ b/pkg/roachpb/data.pb.go
@@ -440,8 +440,10 @@ type Transaction struct {
 	// clock, we may add that to the map, which eliminates read uncertainty for
 	// reads on that node.
 	//
-	// The list of observed timestamps is kept sorted by NodeID. Use
-	// Transaction.UpdateObservedTimestamp to maintain the sorted order.
+	// The slice of observed timestamps is kept sorted by NodeID. Use
+	// Transaction.UpdateObservedTimestamp to maintain the sorted order. The
+	// slice should be treated as immutable and all updates should be performed
+	// on a copy of the slice.
 	ObservedTimestamps []ObservedTimestamp `protobuf:"bytes,8,rep,name=observed_timestamps,json=observedTimestamps" json:"observed_timestamps"`
 	// Writing is true if the transaction has previously sent a Begin transaction
 	// (i.e. if it ever attempted to perform a write, so if it ever attempted to
@@ -463,8 +465,10 @@ type Transaction struct {
 	// the transaction contains any calls to DeleteRange, in order to
 	// prevent the LostDeleteRange anomaly. This flag is relevant only
 	// for SNAPSHOT transactions.
-	RetryOnPush bool   `protobuf:"varint,13,opt,name=retry_on_push,json=retryOnPush,proto3" json:"retry_on_push,omitempty"`
-	Intents     []Span `protobuf:"bytes,11,rep,name=intents" json:"intents"`
+	RetryOnPush bool `protobuf:"varint,13,opt,name=retry_on_push,json=retryOnPush,proto3" json:"retry_on_push,omitempty"`
+	// The slice should be treated as immutable and all updates should be
+	// performed on a copy of the slice.
+	Intents []Span `protobuf:"bytes,11,rep,name=intents" json:"intents"`
 	// Epoch zero timestamp is used to keep track of the earliest timestamp
 	// that any epoch of the transaction used. This is set only if the
 	// transaction is restarted and the epoch is bumped. It is used during

--- a/pkg/roachpb/data.proto
+++ b/pkg/roachpb/data.proto
@@ -351,8 +351,10 @@ message Transaction {
   // clock, we may add that to the map, which eliminates read uncertainty for
   // reads on that node.
   //
-  // The list of observed timestamps is kept sorted by NodeID. Use
-  // Transaction.UpdateObservedTimestamp to maintain the sorted order.
+  // The slice of observed timestamps is kept sorted by NodeID. Use
+  // Transaction.UpdateObservedTimestamp to maintain the sorted order. The
+  // slice should be treated as immutable and all updates should be performed
+  // on a copy of the slice.
   repeated ObservedTimestamp observed_timestamps = 8 [(gogoproto.nullable) = false];
   // Writing is true if the transaction has previously sent a Begin transaction
   // (i.e. if it ever attempted to perform a write, so if it ever attempted to
@@ -375,6 +377,8 @@ message Transaction {
   // prevent the LostDeleteRange anomaly. This flag is relevant only
   // for SNAPSHOT transactions.
   bool retry_on_push = 13;
+  // The slice should be treated as immutable and all updates should be
+  // performed on a copy of the slice.
   repeated Span intents = 11 [(gogoproto.nullable) = false];
   // Epoch zero timestamp is used to keep track of the earliest timestamp
   // that any epoch of the transaction used. This is set only if the

--- a/pkg/roachpb/data_test.go
+++ b/pkg/roachpb/data_test.go
@@ -550,8 +550,10 @@ func TestTransactionClone(t *testing.T) {
 	// listed below. If this test fails, please update the list below and/or
 	// Transaction.Clone().
 	expFields := []string{
+		"Intents",
 		"Intents.EndKey",
 		"Intents.Key",
+		"ObservedTimestamps",
 		"TxnMeta.Key",
 	}
 	if !reflect.DeepEqual(expFields, fields) {

--- a/pkg/roachpb/errors.go
+++ b/pkg/roachpb/errors.go
@@ -493,7 +493,7 @@ func NewReadWithinUncertaintyIntervalError(
 	if txn != nil {
 		maxTS := txn.MaxTimestamp
 		rwue.MaxTimestamp = &maxTS
-		rwue.ObservedTimestamps = append([]ObservedTimestamp(nil), txn.ObservedTimestamps...)
+		rwue.ObservedTimestamps = txn.ObservedTimestamps
 	}
 	return rwue
 }

--- a/pkg/roachpb/errors.go
+++ b/pkg/roachpb/errors.go
@@ -169,9 +169,9 @@ func (e *Error) GetTxn() *Transaction {
 	return e.UnexposedTxn
 }
 
-// UpdateTxn updates the txn.
+// UpdateTxn updates the error transaction.
 func (e *Error) UpdateTxn(o *Transaction) {
-	if e == nil {
+	if o == nil {
 		return
 	}
 	if e.UnexposedTxn == nil {

--- a/pkg/roachpb/errors.go
+++ b/pkg/roachpb/errors.go
@@ -149,7 +149,6 @@ func (e *Error) GoError() error {
 
 // SetTxn sets the txn and resets the error message. txn is cloned before being
 // stored in the Error.
-// TODO(kaneda): Unexpose this method and make callers use NewErrorWithTxn.
 func (e *Error) SetTxn(txn *Transaction) {
 	e.UnexposedTxn = txn
 	if txn != nil {

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -3069,18 +3069,13 @@ func (s *Store) Send(
 			// this node.
 			if pErr != nil {
 				pErr.OriginNode = ba.Replica.NodeID
-				if txn := pErr.GetTxn(); txn != nil {
-					// Clone the txn, as we'll modify it.
-					pErr.SetTxn(txn)
-				} else {
+				if txn := pErr.GetTxn(); txn == nil {
 					pErr.SetTxn(ba.Txn)
 				}
-				pErr.GetTxn().UpdateObservedTimestamp(ba.Replica.NodeID, now)
 			} else {
 				if br.Txn == nil {
 					br.Txn = ba.Txn
 				}
-				br.Txn.UpdateObservedTimestamp(ba.Replica.NodeID, now)
 				// Update our clock with the outgoing response txn timestamp
 				// (if timestamp has been forwarded).
 				if ba.Timestamp.Less(br.Txn.Timestamp) {

--- a/pkg/testutils/storageutils/mocking.go
+++ b/pkg/testutils/storageutils/mocking.go
@@ -58,11 +58,7 @@ func WrapFilterForReplayProtection(
 func shallowCloneErrorWithTxn(pErr *roachpb.Error) *roachpb.Error {
 	if pErr != nil {
 		pErrCopy := *pErr
-		txn := pErrCopy.GetTxn()
-		if txn != nil {
-			txnClone := txn.Clone()
-			pErrCopy.SetTxn(&txnClone)
-		}
+		pErrCopy.SetTxn(pErrCopy.GetTxn())
 		return &pErrCopy
 	}
 

